EDA/augmentations.py

```diff
@@ -0,0 +1,275 @@
+
+def image_label_pairs(images_dir: Path, labels_dir: Path):
+    pairs = []
+    for img_path in images_dir.glob("*"):
+        if img_path.suffix.lower() not in img_extensions():
+            continue
+        label_path = labels_dir / (img_path.stem + ".txt")
+        pairs.append((img_path, label_path))
+    return pairs
+
+def resolve_path(base_dir: Path, p: str) -> Path:
+    q = Path(p)
+    return q if q.is_absolute() else (base_dir / q)
+
+def main():
+    parser = argparse.ArgumentParser(description="Offline long-tail augmentation for YOLO datasets (Ultralytics layout)")
+    parser.add_argument("--data_yaml", type=str, required=True, help="Path to data.yaml (Ultralytics style)")
+    parser.add_argument("--floor", type=int, default=150, help="Minimum train instances per class after augmentation")
+    parser.add_argument("--max_per_image", type=int, default=3, help="Max augmented copies per source image per pass")
+    parser.add_argument("--suffix", type=str, default="_aug", help="Suffix for augmented files")
+    parser.add_argument("--limit", type=int, default=0, help="Optional cap on total augmented samples to generate (0=unlimited)")
+    parser.add_argument("--seed", type=int, default=42, help="Random seed")
+    args = parser.parse_args()
+
+    random.seed(args.seed)
+    np.random.seed(args.seed)
+
+    data_yaml_path = Path(args.data_yaml).resolve()
+    data = load_yaml(data_yaml_path)
+    base_dir = data_yaml_path.parent
+
+    train_spec = str(data["train"])
+    train_path = resolve_path(base_dir, train_spec)
+
+    if train_path.name.lower() == "images":
+        images_dir = train_path
+        labels_dir = train_path.parent / "labels"
+    else:
+        images_dir = train_path / "images"
+        labels_dir = train_path / "labels"
+
+    if not images_dir.exists() or not labels_dir.exists():
+        raise FileNotFoundError(f"Could not find expected train/images and train/labels under {train_path} "
+                                f"(resolved from '{train_spec}' relative to {base_dir})")
+
+    print(f"[INFO] Using images: {images_dir}")
+    print(f"[INFO] Using labels:  {labels_dir}")
+
+    counts = compute_class_counts(labels_dir)
+    if not counts:
+        print("[WARN] No labels found in train/labels. Nothing to augment.")
+        return
+
+    all_classes = sorted(counts.keys())
+    print("\n[BEFORE] Train instances per class:")
+    for c in all_classes:
+        print(f"  class {c}: {counts[c]}")
+
+    pairs = image_label_pairs(images_dir, labels_dir)
+    class_to_images = defaultdict(list)
+    for idx, (img_path, lbl_path) in enumerate(pairs):
+        boxes = read_yolo_labels(lbl_path)
+        if not boxes:
+            continue
+        for c in set(b[0] for b in boxes):
+            class_to_images[c].append(idx)
+
+    augment = build_augment_pipeline()
+
+    total_created = 0
+    progress = True
+    pbar = tqdm(total=1, desc="Augmenting", dynamic_ncols=True)
+    while progress:
```

Comment:
you can use something like huggingface datasets to do more parralel processing and it has some nice calls already